Fix Iceberg Reader for nested partitions (#575)

andrei-ionescu opened the pull request:
This is a fix for #575 — **After writing Iceberg dataset with nested partitions it cannot be read anymore**.

andrei-ionescu commented:
cc @rdblue @aokolnychyi @fbocse @prodeezy @rominparekh

rdsr commented on spark/src/test/java/org/apache/iceberg/spark/source/TestParquetWrite.java:
```diff
@@ -338,4 +339,65 @@ public void testPartitionedCreateWithTargetFileSizeViaOption() throws IOExceptio
+    HadoopTables tables = new HadoopTables(new Configuration());
+    PartitionSpec spec = PartitionSpec.builderFor(nestedSchema)
+        .identity("id")
+        .identity("nestedData.moreData")
```
I'm wondering whether whether we should also fix for partition field names and not allow dots in them. Partition spec schema is used in `DataFile` schema which is used by the `ManifestReader` . 
Iceberg schemas support `indexByName` api  which maps a source column name to its field id. The column name is canonical, starting from root and containing all parent field names separated by dot.

@rdblue do you think this could somehow break the Schema `indexbyName` api?

andrei-ionescu replied:
This dot issue in column names is fixed on the writer side by:

- Issue: https://github.com/apache/incubator-iceberg/issues/575
- PR: https://github.com/apache/incubator-iceberg/pull/220

I would say that to be consistent we should unify this on both writer and reader - that is what this PR tries to do - if we allow dots on write the reader should also work.

rdblue replied:
Dots are allowed in all names. This is possible because we look up columns using indexByName, which has the nested representation of all field names, like "location.lat".

To build the lookup index, field names in the schema are converted once to the fully-nested name. For example, `{ "location": {"lat": ..., "long": ...}, "id" }` has 3 leaf names: location.lat, location.long, and id. Note that "location.lat" could be a top-level field with a dot in the name. That's allowed as long as there isn't a conflicting nested field. If there were a conflict, then "location.lat" would refer to two fields and building the index fails.

Using this index, we can always get the original field ID without ever needing to parse a field name using dot and traverse levels.

rdsr commented on core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java:
```diff
@@ -81,13 +81,14 @@ public Schema record(Schema record, List<String> names, Iterable<Schema.Field> s
     List<Types.NestedField> expectedFields = struct.fields();
     for (int i = 0; i < expectedFields.size(); i += 1) {
       Types.NestedField field = expectedFields.get(i);
+      String sanitizedFieldName = AvroSchemaUtil.sanitize(field.name());
```
When converting an Iceberg schema to an Avro schema if ever a field was sanitized, the original field name was stored as a field property under `AvroSchemaUtil.ICEBERG_FIELD_NAME_PROP` . Should we use that original name if it was set, instead of sanitizing all Iceberg fields before comparison.

andrei-ionescu replied:
I tried your suggestion by getting the original name of the `field` by using `field.getProp(AvroSchemaUtil.ICEBERG_FIELD_NAME_PROP)` — the `expectedField` doesn't have it. Method `getProp` return null if the field has not been sanitized as the `ICEBERG_FIELD_NAME_PROP` is not set. This leads to more checks in all conditions bellow.

First, I guess we will still have to check if the `name`s are different for the cases where no sanitization is applied.
Second, we will have to check if the `ICEBERG_FIELD_NAME_PROP` is null itself.

Now, if the `AvroSchemaUtil.sanitize` is a pure function and the sanitization cases are correctly defined and desired to be applied everywhere  a field is `!validAvroName`, there should not be any issues using it in other parts over original fields — the output should be the same.

What do you think? Should I change this even though the code will be less readable due to checks?

xabriel commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,7 +397,16 @@ public void close() throws IOException {
       PartitionSpec spec = task.spec();
-      Set<Integer> idColumns = spec.identitySourceIds();
+
+      // exclude nested partitions as the data already contains the nested field values
```
Did you mean "exclude nested columns"?

andrei-ionescu replied:
I'll remove the comment.

xabriel commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,7 +397,16 @@ public void close() throws IOException {
-      Set<Integer> idColumns = spec.identitySourceIds();
+
+      // exclude nested partitions as the data already contains the nested field values
+      Set<Integer> idColumns = new HashSet<>();
```
Prefer `Sets.newHashSet();`

andrei-ionescu replied:
👍

rdblue commented:
I don't think that this fix is correct. It looks like this allows creating a projection schema with a `.` in the name, but that isn't actually returning the data as it would be.

Say I have records like this: `{"id": 1, "info": {"type": "x", ...} }` and I want to store those records by `identity(info.type)` and `bucket(id, 16)`. When I read the data, I should get the original record back, with "type" nested within "info". But this approach instead would return `{"id": 1, "info.type": "x", "type": ...}`. Right?

I think the solution is to update the Spark code to use constant readers, like we recently added to Pig.

The current Spark implementation creates a join row of partition data that must be flat, and then combines it with a row that is materialized from the data file. Instead, we need to materialize the entire record, but instead of reading the partition values, we want to set them to constants instead. That's what the ConstantReader in Pig does. Instead of performing a read from the file to get the data, it just returns the known partition value. This works to fix the current problem because the constant reader is at the right place in the value reader tree, so it looks like this:

```
StructReader(
  LongReader(pos=0, name="id", decoder=...),
  StructReader(name="info",
    ConstantReader(pos=0, name="type", value="x"),
    StringReader(pos=1, decoder=...),
    ...
  )
);
```

rdblue commented:
Here's where the constant readers were added to Pig: https://github.com/apache/incubator-iceberg/commit/d158818cb90beed3298be48a0a9de2366ba3cdb5

That approach should work for Spark as well.

prodeezy commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,7 +397,15 @@ public void close() throws IOException {
       PartitionSpec spec = task.spec();
-      Set<Integer> idColumns = spec.identitySourceIds();
+
+      Set<Integer> idColumns = Sets.newHashSet();
```
can you check if this adds latency for very large schemas? Since this is order of magnitude of the size of schema.

prodeezy commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,7 +397,15 @@ public void close() throws IOException {
-      Set<Integer> idColumns = spec.identitySourceIds();
+
+      Set<Integer> idColumns = Sets.newHashSet();
+      for (Integer i : spec.identitySourceIds()) {
```
can you add a comment on why this check is needed?

andrei-ionescu replied:
With this we check the nested partitions and we removed them from the right join side. We do this because the the left join already has the nested values inside because that spark doesn't remove remove nested fields even though you specify `partitionBy("nested.field")`.

So, in the case of doing `partitionBy("nested.field")` on a simple spark data frame and writing it as parquet, the result is thee fact that the file paths contains `.../nested.field=<value>/...` while the nested field is still present inside the data itself (inside the `.parquet` file).

rdblue replied:
I think I see. Sounds like you're decoding the data written in data files, which keeps the existing structure. Is that right?

I'd still prefer to use the `ConstantReader` approach. Iceberg writes partition values to data files, but files imported from Hive tables -- which we do support -- won't have those values and those columns will be missing from the final data.

andrei-ionescu replied:
@rdblue I'll try use a unit test over Hive to see the issue you are describing.

On the other hand I tried joining nested fields using the `JoinedRow` construct but it didn't suffice. 

Given the following original row:
```json
{
  "id": 1,
  "otherField": "otherVal", 
  "nestedData": { 
    "nestedKey1": "val1",  
    "nestedKey2": "val2" 
  } 
}
```
And writing it as Iceberg with the following 2 partitions: `identity(id)` and `identity(nestedData.nestedKey1)`.

Reading it back with Iceberg Reader would mean creating the row as a join between the partition values (ie: `1` and `val1`) and data (ie: `otherVal` and `val2`).

So, joining the following partition values
```json
{ 
  "id": 1, 
  "nestedData": {
    "nestedKey1": "val1" 
  } 
}
```
and data
```json
{ 
  "otherField": "otherVal", 
  "nestedData": { 
    "nestedKey2": "val2" 
  }
}
```
would end as 
```json
{ 
  "id": 1, 
  "nestedData": { 
    "nestedKey1": "val1" 
  },  
  "otherField": "otherVal", 
  "nestedData": { 
    "nestedKey2": "val2" 
  } 
}
```
The above is from what I've seen while putting together this fix.

But we would like it like this:
```json
{ 
  "id": 1, 
  "otherField": "otherVal", 
  "nestedData": { 
    "nestedKey1": "val1",  
    "nestedKey2": "val2" 
  } 
}
```
Right?

I don't know how the `ConstantReader` would help composing back that `nestedData` field as it originally was.

andrei-ionescu replied:
> Iceberg writes partition values to data files, but files imported from Hive tables -- which we do support -- won't have those values and those columns will be missing from the final data

@rdblue can you point me to any unit test or snippet of code showing that flow -- loading Hive table data? I'm asking so I can have a starting point as my experience with Hive is not at the same level with Parquet. Thank you.

rdblue replied:
Sorry for the late reply!

Partitioned tables converted using the [`importSparkTable`](https://github.com/apache/incubator-iceberg/blob/master/spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala#L365-L368) won't have partition columns embedded in the file.

But that's not the only reason to use the `ConstantReader` approach. As I mentioned, using a `ConstantReader` will allow us to produce the schema that is requested by Spark without using a `JoinedRow` and a projection. The projection is expensive, so fixing this using a `ConstantReader` and removing the projection is a much better solution.

andrei-ionescu commented:
@rdblue If you look at this line (https://github.com/apache/incubator-iceberg/pull/585/files#diff-e6baf84fa0dea08f3457433c123f784fR393) in the test  added you'll see that the partition is not flattened as you expected.

Given your example the output would be: `{"id": 1, "info": {"type": "x"} }`. The test checks exactly this - make sure that we keep the nesting as it is.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java:
```diff
@@ -81,13 +81,14 @@ public Schema record(Schema record, List<String> names, Iterable<Schema.Field> s
 
       // detect reordering
-      if (i < fields.size() && !field.name().equals(fields.get(i).name())) {
+      if (i < fields.size() && !sanitizedFieldName.equals(fields.get(i).name())) {
```
Why is sanitization necessary if this fix works by removing columns from the partition columns that get attached via `JoinedRow`?

andrei-ionescu replied:
This is happening before getting to the materialized data. Matching the columns at read time did totally lack the sanitization logic that is used at write time.

rdblue commented:
@andrei-ionescu, let me know if you need any assistance with this. I think it will be great to get this fixed and remove the hacky unsafe projections in Spark!

andrei-ionescu commented:
@rdblue I'll try to change the code so that it won't need the unsafe projection anymore. I'll try have that as a separate commit. If I get stuck I'll get back to you. Thanks for support.

andrei-ionescu commented:
@rdblue I added a new commit with the implementation without the unsafe projection. Can you check it please?

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -19,11 +19,13 @@
 
 package org.apache.iceberg.spark.source;
 
+import com.google.common.base.Function;
```
We use `java.util.function` classes instead of the Guava ones.

andrei-ionescu replied:
👍

andrei-ionescu commented:
@rdblue I modified the code and used a lambda function instead of the Guava version. I also rebased it and resolved conflicts. Please have another look when you have some spare time. Thank you.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,45 +396,57 @@ public void close() throws IOException {
+              if (ref > internalRow.numFields()) {
+                internalRow.setNullAt(ref);
+              }
+              internalRow.update(ref, ((GenericInternalRow) partition).genericGet(index));
```
Looks like what you're trying to do is match up the partition fields with fields in the schema, using the field index for data coming from partition and the source ID to recover an index into the output row.

Unfortunately, there are several problems with this approach. The biggest one is that the source ID can't be used to produce an index. This may appear to work when all source fields are projected, but if you don't project a field, then that index is off by 1. It also assumes that IDs are assigned a certain way and there is not guarantee for that in the format. Reordering fields also causes problems with this approach.

The problems can be fixed by using the approach I recommended earlier: this should build a map of source ID to partition value. Then it should pass that map in when building the [reader tree](https://github.com/apache/incubator-iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/data/SparkParquetReaders.java#L70), so that the tree can use constant readers instead of primitive readers. Here's how it happens in Pig: https://github.com/apache/incubator-iceberg/blob/master/pig/src/main/java/org/apache/iceberg/pig/PigParquetReader.java#L156-L160

andrei-ionescu commented:
@rdblue I think I understood what you wanted. Please check and let me know if it's going in the right direction.

rdblue commented:
@andrei-ionescu, thanks for the update! I'll review it in the next couple days.

rdblue commented:
The current direction of the PR looks like the right one to me. Thanks!

rdblue commented on spark/src/jmh/java/org/apache/iceberg/spark/data/parquet/SparkParquetReadersFlatDataBenchmark.java:
```diff
@@ -116,7 +117,7 @@ public void tearDownBenchmark() {
     try (CloseableIterable<InternalRow> rows = Parquet.read(Files.localInput(dataFile))
         .project(SCHEMA)
-        .createReaderFunc(type -> SparkParquetReaders.buildReader(SCHEMA, type))
+        .createReaderFunc(type -> SparkParquetReaders.buildReader(SCHEMA, type, Collections.emptyMap()))
```
Can we add a `buildReader` method that defaults the map to empty? Then we wouldn't need to change this file at all.

andrei-ionescu replied:
👍

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,48 +397,48 @@ public void close() throws IOException {
         Schema partitionSchema = TypeUtil.select(requiredSchema, idColumns);
         PartitionRowConverter convertToRow = new PartitionRowConverter(partitionSchema, spec);
-        JoinedRow joined = new JoinedRow();
+        GenericInternalRow partition = (GenericInternalRow) convertToRow.apply(file.partition());
```
I'd rather have this go directly from `file.partition()` to the partition map. The problem with using the old partition row is that this has to account for the row's schema as well as the partition schema. It would be easier to avoid this and use a conversion function for values instead of relying on `PartitionRowConverter` to do it.

andrei-ionescu replied:
👍

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,48 +397,48 @@ public void close() throws IOException {
+        Map<Integer, Object> partitionValueMap = Maps.newHashMap();
+        Map<String, Integer> partitionSpecFieldIndexMap = Maps.newHashMap();
+        for (int i = 0; i < spec.fields().size(); i++) {
+          partitionSpecFieldIndexMap.put(spec.fields().get(i).name(), i);
```
These indexes are the position in the spec, not position in the partition row.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,48 +397,48 @@ public void close() throws IOException {
-        iter = Iterators.transform(open(task, readSchema), joined::withLeft);
+        for (Types.NestedField field : partitionSchema.columns()) {
+          int partitionIndex = partitionSpecFieldIndexMap.get(field.name());
+          partitionValueMap.put(field.fieldId(), partition.genericGet(partitionIndex));
```
This get uses the wrong index. It should be `convertPartitionValue(file.partition().get(partitionIndex, Object.class))` instead.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,48 +397,48 @@ public void close() throws IOException {
-      // TODO: remove the projection by reporting the iterator's schema back to Spark
-      return Iterators.transform(iter,
-          APPLY_PROJECTION.bind(projection(finalSchema, iterSchema))::invoke);
+      return iter;
```
I don't think we should remove the projection in this PR. There is still the case where the `requiredSchema` and the `finalSchema` are not the same.

I don't think that code path is ever actually used because Spark requests all columns needed to run filters and Iceberg tells Spark to run all of the filters (because Spark can't handle residuals per task). But to remove this, we need to also remove the `hasExtraFilterColumns` code path, which would mean adding an unrelated change to this PR and as well as adding more tests. So let's separate those into different changes and leave the projection here.

andrei-ionescu commented:
@rdblue I adjusted the code and includes the following main changes:

1. Get rid of `PartitionRowConverter`
2. Add back the projection

I would need you to have a close look on the following two sections that I'm not sure:

1. Column sorting: https://github.com/apache/incubator-iceberg/pull/585/files#diff-a1987550582f01b2aaaff011076bcf1dR442
2. Partition type retrieval: https://github.com/apache/incubator-iceberg/pull/585/files#diff-a1987550582f01b2aaaff011076bcf1dR556

I renamed the commit as we no more are getting rid of the projection.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,48 +397,64 @@ public void close() throws IOException {
-      StructType sparkType = SparkSchemaUtil.convert(finalSchema);
-      Schema requiredSchema = SparkSchemaUtil.prune(tableSchema, sparkType, task.residual(), caseSensitive);
+      Schema requiredSchema = SparkSchemaUtil.prune(tableSchema,
+          SparkSchemaUtil.convert(finalSchema), task.residual(), caseSensitive);
```
This change doesn't look necessary.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -533,69 +552,27 @@ private static UnsafeProjection projection(Schema finalSchema, Schema readSchema
-          reusedRow.update(i, convert(value, types[i]));
-        } else {
-          reusedRow.setNullAt(i);
+    private Object getPartitionValue(StructLike schema, int index, DataType sparkType, Class<?> javaType) {
```
Can you restore the original `convert` method that was removed and just add support for null values to it? That would reduce the amount of churn in this PR.

andrei-ionescu replied:
Adding back the `convert` method implies adding back the full `PartitionRowConverter` class. Are you ok with this?

rdblue replied:
No, that's a static method. We can just move it out of the PartitionRowConverter class.

andrei-ionescu replied:
Ok. Got it. :+1:

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -504,12 +521,14 @@ private static UnsafeProjection projection(Schema finalSchema, Schema readSchema
           .project(readSchema)
           .split(task.start(), task.length())
-          .createReaderFunc(fileSchema -> SparkParquetReaders.buildReader(readSchema, fileSchema))
+          .createReaderFunc(fileSchema -> SparkParquetReaders.buildReader(readSchema, fileSchema, partitionValues))
```
Does Avro need to be updated as well?

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,48 +397,64 @@ public void close() throws IOException {
-        joined.withRight(partition);
+        Map<Integer, Object> partitionValueMap = Maps.newHashMap();
+        Map<String, Integer> partitionSpecFieldIndexMap = Maps.newHashMap();
+        Map<String, Class<?>> partitionSpecJavaTypeMap = Maps.newHashMap();
```
Why are these indexed by name instead of by ID?

andrei-ionescu replied:
I’ll try to switch to ID indexing. It was simpler to use names.

rdblue replied:
Also, I don't think you need to use the Java type. You should be able to pass Object.class instead since the conversion code will cast.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,48 +397,64 @@ public void close() throws IOException {
+          columns.add(field);
+        }
 
+        columns.sort(Comparator.comparingInt(Types.NestedField::fieldId));
```
Columns should not be sorted. What was the rationale for adding this?

andrei-ionescu replied:
The partition columns are added at the end of  the schema thus leading to cast errors inside the projection dynamic code.

rdblue replied:
I think that's probably caused by using the wrong schema to build the case. You should be building data with the required schema, then casting to the final schema.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -396,48 +397,64 @@ public void close() throws IOException {
+        ) {
+          idColumns.add(i);
+        }
+      }
```
Why is this still running a check on identity partitions? The partition value map should include all identity partition values and they will be added in the right place. Fields that aren't projected will be ignored.

andrei-ionescu commented:
Hey @rdblue, I simplified the reader implementation taking into account all your suggestions.

Listing a few of changes:

- removed the java types map and used `Object.class`
- added back the `convert` method with support for `null`
- switched to field id indexing instead of field name
- removed the column sort/ordering

Please have another look when you have time.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -588,15 +564,18 @@ public InternalRow apply(StructLike tuple) {
       }
       return value;
     }
+
```
Nit: non-functional whitespace change.

andrei-ionescu replied:
👍

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -533,51 +538,22 @@ private static UnsafeProjection projection(Schema finalSchema, Schema readSchema
+    /**
+     * Returns the Spark data type for a partition.
+     *
+     * If the field is not found in the partition schema we try to retrieve it from the full table schema.
```
If the field is not in the partition schema, I think this should throw an exception because something is wrong.

andrei-ionescu replied:
This will be removed after I switch from `partitionSchema` to `spec.partitionType()`.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -408,36 +405,42 @@ public void close() throws IOException {
+        for (int i = 0; i < spec.fields().size(); i++) {
+          Integer sourceId = spec.fields().get(i).sourceId();
+          partitionSpecFieldIndexMap.put(sourceId, i);
+          partitionSpecDataTypeMap.put(sourceId, getPartitionType(partitionSchema, sourceId));
```
Why does this use `partitionSchema` instead of `spec.partitionType()`? The partition schema is made from the partition columns in the table schema, which could be out of order. The spec's type is the struct type returned by `file.partition()`.

andrei-ionescu replied:
👍

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -398,7 +396,6 @@ public void close() throws IOException {
       PartitionSpec spec = task.spec();
       Set<Integer> idColumns = spec.identitySourceIds();
 
-      // schema needed for the projection and filtering
```
Why remove this comment?

andrei-ionescu replied:
Added back.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -408,36 +405,42 @@ public void close() throws IOException {
-        // create joined rows and project from the joined schema to the final schema
-        iterSchema = TypeUtil.join(readSchema, partitionSchema);
-        iter = Iterators.transform(open(task, readSchema), joined::withLeft);
+        Set<Integer> projectedIds = TypeUtil.getProjectedIds(partitionSchema);
```
Why use projected IDs? This could just make a map of all partition values instead.

andrei-ionescu replied:
👍

andrei-ionescu commented:
Rebased, and integrated feedback. @rdblue please have one more look.

andrei-ionescu commented:
@rdblue, can you have a look into this once more? All feedback is addressed. Thank you.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -408,36 +406,39 @@ public void close() throws IOException {
+        }
 
+        iterSchema = requiredSchema;
+        iter = open(task, finalSchema, partitionValueMap);
```
This isn't right. This reports that the iterator's schema (`iterSchema`) is `requiredSchema`, but then opens tasks with `finalSchema`.

rdblue commented:
@andrei-ionescu, I opened https://github.com/andrei-ionescu/incubator-iceberg/pull/1 with the fixes needed for the Parquet path in this PR.

Unfortunately, this also changes what the Spark reader expects from data formats, so we will either need to handle Parquet separately (with the changes from my PR) or update Avro to add constant readers as well. Updating Avro is quite a bit harder, so I think it makes sense to separate handling for Parquet and use the existing join approach for Avro data.

andrei-ionescu commented:
@rdblue, I merged your PR, and added `day` partition into the nested partitions related unit test.

Let me know how I can help.

rdblue commented:
@andrei-ionescu, to get this in, we need the updated logic to be applied only to Parquet. Avro and ORC should still use the joined row approach because they have not been updated. Can you make that change?

andrei-ionescu commented:
@rdblue, are there any tests for these cases? Or should I add ones? If so, can you point me to a starting point?

rdblue commented:
I think we need new tests.

I'd start with a test case for each format that writes a few simple records with identity-partitioned columns. The tests should be able to add a partition column that isn't in the file and should validate that data from the file partition tuple overrides what's in the file (if it is a different value).

andrei-ionescu commented:
@rdblue, 

I added a nested partition test for Avro similar to the one for Parquet. Let me know if it's enough or not.

For Orc format, the writer doesn't support it as seen in [Writer.java#L289-L308](https://github.com/apache/incubator-iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/source/Writer.java#L289-L308). Do you want me to try addressing this missing Orc writer feature in a separate PR?

andrei-ionescu commented:
@rdblue Please don't forget about this PR. Thank you.

rdblue commented:
@andrei-ionescu, this hasn't been updated to handle Avro and Parquet differently. This changes Parquet to use partition values from each DataFile, but it does not update Avro (or ORC). That's why Avro needs to use the same logic as before, with a joined row. This should also test that reading a data file without the partition columns works for both Avro and Parquet.

andrei-ionescu commented:
@rdblue, 

I don’t fully understand what you mean and what you’re asking. I added a test that uses Avro as the underlying store format and it works very well without any changes. What’s the case/test that require the change back to joining rows?

I’ll add the test for Avro without any partitions as you suggest.

I guess I’m missing something that you are aware of and that is why I need more information.

rdblue commented:
@andrei-ionescu, this PR changes the way partition values are added to result records for Parquet. It does not change Avro or ORC. That means that the changes in Reader that remove the `JoinedRow` should not be applied to Avro or ORC. This PR needs to make changes just to the Parquet read, not for Avro or ORC.

aokolnychyi commented:
I did not follow this thread but wanted to mention [this](https://github.com/apache/spark/pull/26929) Spark PR as it might be relevant.

rdblue commented:
@aokolnychyi, it isn't clear if that Spark PR is attempting to disable this for just sources that don't support it, or for all sources. I don't think there is a reason to disallow it for all sources. The problem here is just record materialization, which we can fix. The problem on that PR is that some sources don't work and should have checks to reject partitioning by nested columns. There's also a note about HiveQL syntax, but explicitly using partitions in a query is not recommended for Iceberg -- users should instead overwrite by an expression in terms of the data columns.

rdblue commented:
@andrei-ionescu, I just updated the Avro read path with the same change that you're making here in #896. The only outstanding task on this PR was to avoid breaking the other formats (Avro and ORC) and that is done by my PR here: https://github.com/apache/incubator-iceberg/pull/896/files#diff-b9074b510418ffff8aebbaaa95685541R104.

After that goes into master, could you update this PR by adding Parquet to the `SUPPORTS_CONSTANTS` format set?

rdblue commented:
I'm closing this one in because I've made the Parquet changes in #909 and included @andrei-ionescu as a co-author.
